src/main/java/net/imglib2/roi/Bounds.java

```diff
@@ -801,88 +813,74 @@ public void realMinMax( double[] realMin, double[] realMax )
 
 		// -- Helper methods --
 
-		private boolean updateNeeded()
+		private void updateMinMaxIfNeeded()
 		{
-			// TODO (TP): use getMinMax(source, sourceMin[], sourceMax[]) instead
-			//   (with new double[] sourceMin, new double[] sourceMax)
-			//   then reuse sourceMin, sourceMax in checks below and also in updateMinMax().
-			for ( int d = 0; d < transformToSource.numTargetDimensions(); d++ )
+			getMinMax( source, currentSourceMin, currentSourceMax );
+
+			for ( int d = 0; d < numSourceDimensions; d++ )
 			{
-				if ( cachedSourceMin[ d ] != source.realMin( d ) || cachedSourceMax[ d ] != source.realMax( d ) )
-					return true;
+				if ( cachedSourceMin[ d ] != currentSourceMin[ d ] || cachedSourceMax[ d ] != currentSourceMax[ d ] )
+				{
+					System.arraycopy( currentSourceMin, 0, cachedSourceMax, 0 , numSourceDimensions );
+					System.arraycopy( currentSourceMax, 0, cachedSourceMin, 0 , numSourceDimensions );
+					updateMinMax();
+					break;
+				}
 			}
-			return false;
 		}
 
-		private void updateMinMax()
+		private void updateMinMax( )
 		{
 			if( Intervals.isEmpty( source ) )
 			{
 				Arrays.fill( max, Double.NEGATIVE_INFINITY );
 				Arrays.fill( min, Double.POSITIVE_INFINITY );
 			}
-			final double[] sMx = new double[ transformToSource.numTargetDimensions() ];
-			final double[] sMn = new double[ sMx.length ];
 
-			while ( !Arrays.equals( sMx, cachedSourceMax ) ||
-					!Arrays.equals( sMn, cachedSourceMin ) )
-			{
-				source.realMax( sMx );
-				source.realMin( sMn );
+			final double[][] transformedCorners = createCorners();
+			final int numTransformedCorners = transformedCorners.length;
 
-				final double[][] transformedCorners = createCorners();
-				final int numTransformedCorners = transformedCorners.length;
-
-				for ( int d = 0; d < n; d++ )
+			for ( int d = 0; d < n; d++ )
+			{
+				double maxCorner = transformedCorners[ 0 ][ d ];
+				double minCorner = transformedCorners[ 0 ][ d ];
+				for ( int i = 1; i < numTransformedCorners; i++ )
 				{
-					double mx = transformedCorners[ 0 ][ d ];
-					double mn = transformedCorners[ 0 ][ d ];
-					for ( int i = 1; i < numTransformedCorners; i++ )
-					{
-						if ( transformedCorners[ i ][ d ] > mx )
-							mx = transformedCorners[ i ][ d ];
-						if ( transformedCorners[ i ][ d ] < mn )
-							mn = transformedCorners[ i ][ d ];
-					}
-					min[ d ] = mn;
-					max[ d ] = mx;
+					if ( transformedCorners[ i ][ d ] < minCorner )
+						minCorner = transformedCorners[ i ][ d ];
+					if ( transformedCorners[ i ][ d ] > maxCorner )
+						maxCorner = transformedCorners[ i ][ d ];
 				}
-
-				source.realMax( cachedSourceMax );
-				source.realMin( cachedSourceMin );
+				min[ d ] = minCorner;
+				max[ d ] = maxCorner;
 			}
 		}
 
 		private double[][] createCorners()
```

Comment:
@tpietzsch 
I don't understand the code here. If the code is good, shouldn't we put it as a static method somewhere? Something like `RealInterval ... = TransformHelper.estimateBounds(   RealTransform ..., RealInterval ... );`

Reply:
> shouldn't we put it as a static method somewhere

I'd like that, as well as an implementation that would work for non-linear transformations too
https://github.com/imglib/imglib2-realtransform/pull/37

Reply:
I refactored this a bit, and it should be clearer now what happens.

This method computes the corners of the source interval (see javadoc for explanation how)
https://github.com/imglib/imglib2-roi/blob/46e88901c684215b7da82cdd3518b0c56ba96e67/src/main/java/net/imglib2/roi/Bounds.java#L873-L892

Then this method takes the corners and puts them through the transformation
https://github.com/imglib/imglib2-roi/blob/46e88901c684215b7da82cdd3518b0c56ba96e67/src/main/java/net/imglib2/roi/Bounds.java#L858-L865

Finally, this method puts computes the bounding box of the transformed corners
https://github.com/imglib/imglib2-roi/blob/46e88901c684215b7da82cdd3518b0c56ba96e67/src/main/java/net/imglib2/roi/Bounds.java#L841-L855